nexus/db-queries/src/db/datastore/multicast/members.rs

```diff
@@ -0,0 +1,2428 @@
+//! Multicast group member management operations.
+//!
+//! This module provides database operations for managing multicast group memberships,
+//! including adding/removing members and coordinating with saga operations.
+
+use async_bb8_diesel::AsyncRunQueryDsl;
+use chrono::Utc;
+use diesel::prelude::*;
+use slog::debug;
+use uuid::Uuid;
+
+use nexus_db_errors::{ErrorHandler, public_error_from_diesel};
+use omicron_common::api::external::{
+    self, CreateResult, DataPageParams, DeleteResult, ListResultVec,
+    LookupType, ResourceType, UpdateResult,
+};
+use omicron_uuid_kinds::{
+    GenericUuid, InstanceUuid, MulticastGroupUuid, SledKind,
+};
+
+use crate::context::OpContext;
+use crate::db::datastore::DataStore;
+use crate::db::model::{
+    DbTypedUuid, MulticastGroupMember, MulticastGroupMemberState,
+    MulticastGroupMemberValues,
+};
+use crate::db::on_conflict_ext::IncompleteOnConflictExt;
+use crate::db::pagination::paginated;
+
+impl DataStore {
+    /// List members of a multicast group.
+    pub async fn multicast_group_members_list(
+        &self,
+        opctx: &OpContext,
+        group_id: MulticastGroupUuid,
+        pagparams: &DataPageParams<'_, Uuid>,
+    ) -> ListResultVec<MulticastGroupMember> {
+        self.multicast_group_members_list_by_id(
+            opctx,
+            group_id.into_untyped_uuid(),
+            pagparams,
+        )
+        .await
+    }
+
+    /// Get all multicast group memberships for a specific instance.
+    ///
+    /// This method returns all multicast groups that contain the specified
+    /// instance, which is useful for updating multicast membership when
+    /// instances change state.
+    pub async fn multicast_group_members_list_for_instance(
+        &self,
+        opctx: &OpContext,
+        instance_id: Uuid,
+    ) -> ListResultVec<MulticastGroupMember> {
+        use nexus_db_schema::schema::multicast_group_member::dsl;
+
+        diesel::QueryDsl::filter(
+            diesel::QueryDsl::order(
+                diesel::QueryDsl::select(
+                    dsl::multicast_group_member,
+                    MulticastGroupMember::as_select(),
+                ),
+                dsl::id.asc(),
+            ),
+            dsl::parent_id.eq(instance_id).and(dsl::time_deleted.is_null()),
+        )
+        .get_results_async(&*self.pool_connection_authorized(opctx).await?)
+        .await
+        .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// Look up the sled hosting an instance via its active VMM.
+    /// Returns None if the instance exists but has no active VMM
+    /// (stopped instance).
+    pub async fn instance_get_sled_id(
```

Comment:
perhaps this could go in a module with other instance functions since it's not really about multicast membership.

Reply:
Good call.